### PR TITLE
adding grover pretrain model as ModularTorchModel

### PR DESCRIPTION
**Pull request by gusty1g**
# Pull Request Template

## Description

The PR adds GroverPretrainer to train embeddings. The main contribution is the self-supervised pretraining task style pretraining for training embeddings.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:103]: Let's make this layer public. It might be useful for other models. Can you also add it to the docs?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:109]: Can you add docstring?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:120]: Can you use format `(num_bond, feat_dim)`? Where are `num_bond` and `feat_dim` defined? Can you add a definition to the docstring?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:137]: Let's make this public

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:206]: Can you add a docstring with a usage example?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:208]: Can you add docstring/types?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:256]: Can you make the loss a deepchem loss? @tonydavis629 and I have discussed that we should set a new convention that losses should be added with other deepchem losses forr `ModularTorchModel`

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:276]: This function is really complicated; let's discuss offline how we can simplify this implementation

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:349]: Can you add a `Returns` field and a description of the purpose of this method to the docstring?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:368]: Can you add a `Returns` field and a description of the purpose of this method to the docstring?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:406]: The convention we are following in `InfoGraphModel` is to have the same model use a flag to swap between pretraining/fine-tuning. We should follow the same convention here

**Review by rbharath — COMMENTED** [review]
Did a first review pass. This is a good start! Some room for cleanup here. Let's discuss questions offline since there are some design questions open here

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:103]: Fixed in #3277

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:109]: Fixed in #3277

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:120]: Fixed in #3277

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:137]: Fixed in #3277

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:15]: This function is really complicated. Can you add a docstring to explain what it does and provide a usage example? I don't usually ask for usage examples on private functions but it will be very hard to debug this function in future otherwise.

You might also consider breaking this into a few simpler functions

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:67]: We need a docstring and usage example

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:78]: The component names for the model should be fully documented in the docstring for the model

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:190]: Docs here should be converted to numpy style

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:184]: This loss function should be converted to a function and moved to the section of deepchem losses since other models may want to reuse this loss as a general mathematical operation

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:252]: Some debugging cruft here I think

**Review by rbharath — COMMENTED** [review]
I think this PR is still in draft stage and not ready for full review so I just did a quick pass and pointed out a couple things that jumped out at me. Let me know once this is ready for full review

**Review by rbharath — COMMENTED** [review]
Capturing the offline discussion, we will want to split this into multiple PRs:

1. Fixing `GraphData`, `BatchGraphData` to support Grover (and DMPNN)
2. Adding Grover's loss to DeepChem's losses
3. Adding Grover as a modular torch model.

We should also discuss the proposed save/restore fix offline and see if we want to make that a separate PR

**Comment by tonydavis629** [comment]
What is the issue with GraphData not supporting DMPNN and Grover?

File: deepchem/utils/test/test_grover.py
Comment (rbharath) [inline deepchem/utils/test/test_grover.py:3]: As a reminder, we need to add the save/reload/overfit tests before we can merge this PR in

**Review by rbharath — COMMENTED** [review]
I think this PR is waiting on the earlier graph data handling PR so not doing a thorugh review, but a comment that we still need save/reload/overfit tests)

**Review by rbharath — COMMENTED** [review]
There's a lot of code here that I think will go away once we rebase. I'll hold off on doing a full review till we merge in the previous PR and rebase since it'll get messy otherwise

**Comment by gusty1g** [comment]
> What is the issue with GraphData not supporting DMPNN and Grover?

It's not that it does not support - I didn't know how to make DMPNN and Grover graph batching work with GraphData. To add a bit of context, DMPNN and Grover models use same input representations.

**Review by rbharath — COMMENTED** [review]
The core implementation is looking solid, but I have some requests for improved documentation and additional unit tests below.

File: deepchem/models/torch_models/tests/test_grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/tests/test_grover.py:7]: It is a duplicate test. I have moved it to a separate PR (#3329)

**Review by rbharath — COMMENTED** [review]
Will re-review this PR later after we merge the earlier PRs and rebase

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py]: added docs

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:363]: added docstring

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [reply deepchem/models/torch_models/grover.py:264]: Quick reminder to add to the docs

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [reply deepchem/models/torch_models/grover.py:405]: Reminder to add this table.

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [reply deepchem/models/torch_models/grover.py:447]: Reminder here as well

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [reply deepchem/models/torch_models/grover.py:477]: Reminder

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [reply deepchem/models/torch_models/grover.py:490]: Reminder

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [reply deepchem/models/torch_models/grover.py:509]: Reminder

File: deepchem/models/torch_models/tests/test_grover_layers.py
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_grover_layers.py:10]: Do we no longer need this fixture?

**Review by rbharath — COMMENTED** [review]
This is nearly ready but there are still a number of missing docs. Can you do a pass and make sure all comments are addressed?

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:405]: added table

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:264]: added to docs

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:490]: added docstring

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:447]: added description

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:477]: added docs

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:509]: added docstring

File: deepchem/models/torch_models/tests/test_grover_layers.py
Comment (gusty1g) [reply deepchem/models/torch_models/tests/test_grover_layers.py:10]: The fixture was moved to `conftest.py` in #3334 as it is being used be more than one test file.

**Review by rbharath — APPROVED** [review]
LGTM, please go ahead and merge in once CI is clear.

Nice working getting this to the finish line! It's a big one

**Comment by gusty1g** [comment]
The ci failures are not related to these changes. Going ahead and merging this in.

**Pull request merged**
